source/git-api.js

```diff
@@ -31,87 +30,125 @@ exports.registerApi = (env) => {
+    }
+    addGit(item, options) {
+      return this.watchItem('git', item, options);
+    }
+    close() {
+      this.watchers.forEach((w) => w.close());
+    }
+  }
+
+  const readIgnore = async (pathToWatch) => {
+    winston.debug(`Parsing .gitignore for ${pathToWatch}`);
+    const out = ignore();
+    // For Windows/Mac that can't skip
+    out.add('.git/');
+    const ignoreContent = await fs.readFile(path.join(pathToWatch, '.gitignore')).catch(() => null);
+    if (ignoreContent) out.add(ignoreContent.toString());
+    return out;
+  };
+
+  // TODO move to nodegit
+  const watchRepo = async (pathToWatch, ignore) => {
     winston.info(`Start watching ${pathToWatch}`);
-    return fs
-      .access(pathToWatch)
-      .catch(() => {
-        // Sometimes necessary folders, '.../.git/refs/heads' and etc, are not created on git init
-        winston.debug(`intended folder to watch doesn't exists, creating: ${pathToWatch}`);
-        return mkdirp(pathToWatch);
-      })
-      .then(() => {
-        const watcher = watch(pathToWatch, options || {});
-        watcher.on('change', (event, filename) => {
-          if (!filename) return;
-          const filePath = path.join(subfolderPath, filename);
-          winston.debug(`File change: ${filePath}`);
-          if (isFileWatched(filePath, socket.ignore)) {
-            winston.info(`${filePath} triggered refresh for ${socket.watcherPath}`);
-            emitGitDirectoryChanged(socket.watcherPath);
-            emitWorkingTreeChanged(socket.watcherPath);
+    const watcher = new RepoWatcher();
+    let repoPath = path.join(pathToWatch, '.git');
+    let refsPath;
+    if ((await fs.access(repoPath).catch(() => false)) === undefined) {
+      // Looks like a repo, let's watch workdir
+      let gitIgnore = await readIgnore(pathToWatch);
+      await watcher.addWorkdir(pathToWatch, {
+        recursive: true,
+        filter: (changedPath, skip) => {
+          const filePath = path.relative(pathToWatch, changedPath);
+          if (filePath === '.gitignore') {
+            readIgnore(pathToWatch).then(
+              (ign) => (gitIgnore = ign),
+              (err) => winston.error(`Could not parse .gitignore for`, pathToWatch, err)
+            );
           }
-        });
-        watcher.on('error', (err) => {
-          winston.warn(`Error watching ${pathToWatch}: `, JSON.stringify(err));
-        });
-        socket.watcher.push(watcher);
+          if (filePath === '.git') return skip;
+          // We add / to test for directories, we can't have a file named like a directory
+          // and otherwise directory `foo` won't match ignore `foo/`
+          if (gitIgnore.ignores(filePath) || gitIgnore.ignores(`${filePath}/`)) {
+            // TODO https://github.com/kaelzhang/node-ignore/issues/78
+            // optimization: assume these are permanent skips
+            if (/(node_modules|build|dist|cache|coverage)/.test(filePath)) return skip;
```

Comment:
right - that's what issue 78 is about :)

I wonder if it's easy to determine unignores in `ignore`. I have my doubts that issue 78 will see any traction.

Reply:
Also note that this skip only means that changes aren't noticed in the workdir. Git will see them when prodded, so worst case it means reloading the browser window.

I think these names most likely won't have underlying unignores. I'd rather keep this, to improve performance on Linux.

Reply:
Why are we manually skipping all these folders?

I understand that often they should be ignored, but that is a rather dangerous assumption.  

Let's think of this way.  Currently we may miss files users may want to watch if they are in these folders.  If we don't do this skips here, at worst case it just "unoptimized" as watcher will fire and git will scan and see no diffs.

Reply:
well - can we at least do node_modules? Many projects have 1000+ deps which means 1000+ watchers would be attached.

I created https://github.com/kaelzhang/node-ignore/issues/78 in the hope we can do this more correctly.

Whenever you refresh your browser window, it will do a git status which will update the diff view including the workdir. So the worse that can happen is that the user is editing a file in an ignored directory and they have to refresh the window.

Reply:
I would be ok with reducing it to just `node_modules`. 

Pressing the refresh button at the top right does also work:
![image](https://user-images.githubusercontent.com/4009570/150695800-49365bc6-db90-41d1-9f88-637f4bde7f7a.png)

Reply:
I think it's something that shouldn't be changing often and shouldn't trigger this code path but I guess it's fine.